RPP: Dependency Container

RadoslavGeorgiev opened the pull request:
Fixes #6678.

#### Changes proposed in this Pull Request

Introduce a dependency container for the new `src` directory

To achieve that, within this PR:

- The [League container](https://container.thephpleague.com/) is used, as in Core.
- [Auto-Wiring](https://container.thephpleague.com/4.x/auto-wiring/) is intentionally omitted.
- A [Delegate container](https://container.thephpleague.com/4.x/delegate-containers/) allows our container to load WooCommerce core classes.
- A simple base service provider class is implemented, as well as the first service provider for the `PaymentProcessingService`.
- The League Container is extended by `WooPayments\Internal\DependencyManagement\ExtendedContainer` (which adds all necessary additional functionality), while the actual `WooPayments\Container` contains it, and only allows access to selected methods (`get` and `has`).
- [Mozart](https://github.com/coenjacobs/mozart) is used to wrap the League container in our own namespace, through a separate Composer instance.
- The PSR-11 container interface is still included in the root, as it should be useful for compatibility with other plugins.
- The container can be accessed through `wcpay_get_container()`.
- The container is wired up in a way that works in unit tests, allowing classes to be replaced.

#### Docs

The additions in this PR are described in two docs:

1. The readme within `src` ([ref](https://github.com/Automattic/woocommerce-payments/tree/rpp/6678-dependency-container/src)) contains general developed docs.
2. The one in `lib` ([ref](https://github.com/Automattic/woocommerce-payments/tree/rpp/6678-dependency-container/lib)) describes how to manage the container package.
 
#### Testing instructions

Everything is covered in automated tests. Please inspect `tests/unit/src/ContainerTest.php`.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

No applicable post-merge actions.

htdat committed:
Add PaymentService and index files

htdat committed:
Exclude src from WP filename

htdat committed:
Add folder `src` to npm run release

htdat committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Adding the autoloading configuration

htdat committed:
Add changelog

RadoslavGeorgiev committed:
Changelog entry

RadoslavGeorgiev committed:
Removing test left-overs

htdat committed:
Change to PaymentProcessingService, use singular for Service

RadoslavGeorgiev committed:
First changes

RadoslavGeorgiev committed:
Changelog

RadoslavGeorgiev committed:
Update changelog/rpp-6677-psr-autoloader

Co-authored-by: Dat Hoang <htdat@users.noreply.github.com>

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge fix

RadoslavGeorgiev committed:
Switching to a unit test instead of a manual one

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Introducing the `lib` directory.

Because Mozart is a dev dependency, and not available within actions. This means that the lib directory will only be manipulated locally.

RadoslavGeorgiev committed:
Giving lib its own Composer instance

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6862` or branch name `rpp/6678-dependency-container` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=rpp/6678-dependency-container) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b8be247f88b35235f2eca2f95b94bc74d6db67de
- Build time: 2023-08-04 14:13:49 UTC

_Note: the build is updated when a new commit is pushed to this PR._

RadoslavGeorgiev committed:
Making psalm work.

The way we load files for Psalm prevents woocommerce-payments.php from defining functions, which would be otherwise available.

RadoslavGeorgiev committed:
Adding service providers and tests

RadoslavGeorgiev committed:
Tests part 1: Adding `wcpay_get_test_container()`.

RadoslavGeorgiev committed:
Allowing replacements during tests.

RadoslavGeorgiev committed:
Adding the container exception.

RadoslavGeorgiev committed:
Minor cleanups.

RadoslavGeorgiev committed:
Updating the lib readme

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6674-create-payment-service-class

RadoslavGeorgiev committed:
Merge branch 'rpp/6674-create-payment-service-class' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Fixing a unit test mistake

RadoslavGeorgiev committed:
Fixing Psalm issues

RadoslavGeorgiev committed:
Moving the Woo delegate container to a separate directory

RadoslavGeorgiev committed:
Testin improvement, incl. replacements.

Instead of overwriting existing definitions, ExtendedContainer adds a layer of replacements on top of the League container. This way references are kept after reset.

RadoslavGeorgiev commented on woocommerce-payments.php:
```diff
@@ -12,7 +12,7 @@
  * Requires at least: 6.0
  * Requires PHP: 7.3
- * Version: 6.2.0
+ * Version: 6.4.0
```
This was commited by mistake, and will be fixed soon.

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' of github.com:Automattic/woocommerce-payments into rpp/6677-psr-autoloader

RadoslavGeorgiev committed:
Switching namespaces

RadoslavGeorgiev committed:
Merge branch 'rpp/6677-psr-autoloader' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Replacing WooPayments with WCPay

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Docs

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

htdat commented on woocommerce-payments.php:
```diff
@@ -150,6 +149,7 @@ function () {
  */
 function wcpay_init() {
 	require_once WCPAY_ABSPATH . '/includes/class-wc-payments.php';
+	require_once WCPAY_ABSPATH . '/src/wcpay-get-container.php';
```
May config `autoload` in composer?  https://getcomposer.org/doc/04-schema.md#files

Something like this https://gist.github.com/htdat/191438fed64b6d7fefac5e3ff274960d

With this, we do not need to manually load this file here, pslam-loader.php, and tests bootstrap.php

RadoslavGeorgiev replied:
Interesting suggestion. I added it in https://github.com/Automattic/woocommerce-payments/pull/6862/commits/e7e7842549f6b8bb6478deb312f93e17cb52c34b, let's see how it works.

What concerns me a bit is that this way we have less control over the order of what's loaded. On the other hand, nothing should happen there, besides registering a function, so we should be good to go.

It will be interesting what happens once more stuff gets added to `src`, and some of it needs to be loaded directly, instead of through a method in `includes`. In the worst case we'll revert this change, but I have a suspicion that it might end up being helpful instead, forcing us into a more elegant solution.

RadoslavGeorgiev committed:
Switching to 4 spaces within composer.json

RadoslavGeorgiev commented on lib/composer.json:
```diff
@@ -0,0 +1,33 @@
+{
+		"name": "wcpay/wcpay-lib",
+		"description": "A package for hiding re-nameed dependencies.",
```
> Typo: re-nameed

Thanks, that's fixed now.

> Also, it might be good to fix the tabs vs spaces and some other such inconsistency in this and the root composer.json file.

I was actually battling with this in earlier commits, but it got out of focus. Now I switched to 4 spaces for indentation in both files (https://github.com/Automattic/woocommerce-payments/pull/6862/commits/5bf37999d183523864fb9878350b6e4002c5ecde), as Composer uses spaces. A bit more changes in the PR, but at least we'll finally have those files well indented.

RadoslavGeorgiev commented on lib/README.md:
```diff
@@ -0,0 +1,25 @@
+Similarly to how Composer loads packages into the `vendor` directory, this directory will contain WooPayments-specific libraries. There are a couple of differences here:
+
+- All packages have their namespaces prefixed by `WooPayments\Vendor\` to prevent conflicts with other plugins. This is automated through the `coenjacobs/mozart` package.
+- Even though `vendor` packages are not commited to git, the packages within `lib/packages` will be commited.
```
It's a weird experience, trying to wrap a public package within your own namespace.

We need Mozart to achieve this on `composer install` (or `update`), and some builds (ex. live) do not include dev dependencies.

To allow the `lib` package to be installed and updated automatically, we'll need to alter our workflows to first install Composer with all dependencies, then again, but skipping dev dependencies. I was considering the same, but I think it's a complicated and tricky effort, without a significant outcome.

Woo Core has commands for in their root `package.json`, but even those would be just some additional files, which only achieve the same goal. That's why my opinion is that even that's not worth it.

Thoughts?

RadoslavGeorgiev replied:
> I guess the problem could be excluding the "dev" dependencies from the vendor folder after the pull.

Exactly. It's mainly Mozart itself that will need to be excluded, or deleted. Anyway, closing this thread, I think we have a conclusion.

RadoslavGeorgiev commented on src/Container.php:
```diff
@@ -0,0 +1,82 @@
+	public function __construct( WooContainer $woo_container = null ) {
+		$this->container = new ExtendedContainer();
+
+		// Allow the container to be used as a dependency.
+		$this->container->addShared( static::class, $this );
```
> Is this necessary if we already have `wcpay_get_container`?

There's a bit of a mental difference: The container should be used as a dependency within `src`, while `wcpay_get_container()` should be used within `includes`.

>  I would rather the container be used only once at the flow entry point and resolve all possible dependencies, in keeping with inversion of control (the individual classes should not be concerned about how they or others are created, only business logic).

TBH, I was mostly following Woo's example here, and thought it made sense.

In general, I'm 100% 👍 for IoC. However, I don't consider one class being aware of ways to instantiate another one as opposing the principle. Even though DI and IoC go hand in hand, there's a reason why dependencies are not necessarily always shared.

`WCPay\Container` class exists for a reason, and that's to limit the methods, accessible later. Did you notice that only `has` and `get` are available, but others like `add` and `extend` are not? This prevents "mid-level" classes from altering the behavior of the application, so the control remains on top.

There are plenty of cases where it makes sense for a non-shared class to have dependencies, and for those to be provided directly by the container. I've got payment states in my mind right now, wonder who suggested that pattern? 😝 

Anticipating the next question: Technically, we could have payment states as services, and pass the `$payment` object around to all methods. Yet, it also makes sense to instantiate states specifically for a given payment, and using that payment for context, together with all other needed dependencies.

In any case, consider the following: We need the container to be accessible publicly (from `includes`), and as long as `wcpay_get_container()` is accessible publicly, developers will use it even in `src` when needed.

I would rather have the container available as a dependency because that's still better to mock, similarly to a factory:

```php
class Payment {
	private $container;
	private $state;

	public function __construct( Container $container ) {
		$this->container = $container;
	}

	public function process() {
		$state = $this->container->get( SomeState::class );
		$state->set_context( $this );
		$this->state = $state; // I don't 100% like the recursion here by the way.
	}
}

class Payment_Test extends WCPAY_UnitTestCase {
	public function test_payment_transitions_to_state() {
		$mock_container = $this->createMock( Container::class );
		$mock_state     = $this->createMock( SomeState::class );
		
		$mock_container->expects( $this->once() )
			->method( 'get' )
			->with( SomeState::class )
			->willReturn( $mock_state );
		
		$sut = new Payment( $mock_container );

		$mock_state->expects( $this->once() )
			->method( 'set_context' )
			->with( $sut );

		$sut->process();
	}
}
```

---

Another note on a similar topic: Once we migrate server requests to `src`, the case would be very similar for them. Instead of using `WC_Payments::create_request()`, a class would be able to use the container as a dependency, and instantiate requests through it. Those would be mockable in the same way during tests. Of course, requests would need not to be shared.

```php
class SomeService {
	private $container;

	public function __construct( Container $container ) {
		$this->container = $container;
	}

	public function some_method() {
		$request = $this->container->get( SomeRequest::class );
		$request->set_xyz( 123 );
		$request->send();
	}
}
```

I know this is a different can of worms, but it's already open anyway.

RadoslavGeorgiev replied:
> Unrelated, but some comments were left on Zvonimir's recent post about the Requests. Maybe not as a priority, but we might want to revisit them briefly, especially Request instantiation bit.

I have them in mind for some time this week. :)

RadoslavGeorgiev commented on src/Internal/DependencyManagement/ExtendedContainer.php:
```diff
@@ -0,0 +1,102 @@
+/**
+ * Extends the League container to allow WCPay customizations.
+ *
+ * During tests, `wcpay_get_test_container()` will return an instance of this class.
+ */
+class ExtendedContainer extends Container {
```
It is not really exposed to devs because the extended container is a private property of `WCPay\Container`.  `WCPay\Container` does not, and it should never expose those methods of the internal container.

I thought about having a separate container for tests, to avoid shipping those few methods with the plugin, but achieving that would involve hacks. Seems like the easiest and most straight-forward method is to use reflection during tests, and access the private property.

By the way, the main `WCPay\Container` class is still the one that's supplied to classes, and also the one responsible for loading service providers. I don't want to include any code, which allows access to/replacement of the internal `WCPay\Internal\DependencyManagement\ExtendedContainer`.

RadoslavGeorgiev committed:
Fixing an example

RadoslavGeorgiev commented on src/README.md:
```diff
@@ -0,0 +1,180 @@
+		$container->addShared( SimpleService::class );
+
+		$container->addShared( SimpleServiceWithDependencies::class )
+			->addArgument( ServiceWithDependencies::class);
```
Oh, my mistake, thanks for noticing this! Fixed in https://github.com/Automattic/woocommerce-payments/pull/6862/commits/4183481bba1c4739a0ed76e8d68a87114a6ce66f

Note: It will make sense to update this example once `src` has some more real classes.

RadoslavGeorgiev commented on src/README.md:
```diff
@@ -0,0 +1,180 @@
+			->addArgument( ServiceWithDependencies::class);
+
+		$container->add( Request::class )
+			->addArgument( WC_Payments_API_Client::class );
```
The readme will be updated in https://github.com/Automattic/woocommerce-payments/pull/6876, which introduces a `LegacyContainer` that has hard-coded ways to access legacy classes. If the approach there is approved, this example will be valid.

Is there any chance that you could take a quick glance at https://github.com/Automattic/woocommerce-payments/pull/6876/files#diff-97c044d71e0d3b85870e78635d1f2de33e6ba482799e5d61ac21485432ab6f93, and comment on the approach before the PR is polished? Note that the idea is similar to Woo's `LegacyContainer` ([ref](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Proxies/LegacyProxy.php#L69)), but instead of accessing legacy instances through `LegacyContainer`, they would be available as direct dependencies. Once moved to `src`, the classes that depend on them will not require modifications.

RadoslavGeorgiev committed:
Moving the initialization of the container within wcpay_get_container

RadoslavGeorgiev commented on src/wcpay-get-container.php:
```diff
@@ -0,0 +1,19 @@
+ */
+function wcpay_get_container() {
+	return $GLOBALS['wcpay_container'];
+}
```
Done in https://github.com/Automattic/woocommerce-payments/pull/6862/commits/daab07084a284c3ed469e5962efabbedda74a5b8

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container

RadoslavGeorgiev committed:
Use the container for the payment processing service

RadoslavGeorgiev commented on includes/class-wc-payment-gateway-wcpay.php:
```diff
@@ -718,7 +718,7 @@ public function payment_fields() {
 
 		if ( defined( 'WCPAY_NEW_PROCESS' ) && true === WCPAY_NEW_PROCESS ) {
-			$new_process = new PaymentProcessingService();
+			$new_process = wcpay_get_container()->get( PaymentProcessingService::class );
```
Note: The service should be loaded within the constructor of the class, and used later. However, that would require fixing a ton of tests ([ref](https://github.com/Automattic/woocommerce-payments/pull/6421/files)), and this PR already has too many changes.

htdat commented on src/Container.php:
```diff
@@ -0,0 +1,82 @@
+		$this->load_providers();
+
+		// Allow delegating unresolved queries to the WooCommerce container.
+		$this->container->delegate( $woo_container ?? new WooContainer() );
```
Is there any benefit to having a dependency like this `WooContainer $woo_container = null` and then still check if it is `null` and then call itself? 

So we can remove the dependency and make it simple like this one: 
```suggestion
		$this->container->delegate( new WooContainer() );
```

RadoslavGeorgiev replied:
@htdat I did this to allow mocking the container in tests: https://github.com/Automattic/woocommerce-payments/blob/rpp/6678-dependency-container/tests/unit/src/ContainerTest.php#L56-L57

> Is there any benefit to having a dependency like this WooContainer $woo_container = null and then still check if it is null and then call itself?

The line in question simply creates the instance if `$woo_container` isn't already set.

RadoslavGeorgiev replied:
I'm marking this as resolved, and merging the PR, since it's pre-approved.

htdat replied:
Ah, gotcha! If that is for tests, so it makes sense.

htdat commented on src/Container.php:
```diff
@@ -0,0 +1,82 @@
+	/**
+	 * Retrieves an instance of a given class.
+	 *
+	 * @param string $id The ID of the class to retrieve.
+	 * @return mixed
+	 */
```
May use this dynamic return type? That can assist IDEs better. I got that idea from here https://phpstan.org/developing-extensions/dynamic-return-type-extensions and it seems to work well in PhpStorm when I check the change of `class-wc-payment-gateway-wcpay.php` in this PR. 

```suggestion
	 * @template T
	 * @param class-string<T> $id The ID of the class to retrieve.
	 * @return T
	 */
```

htdat replied:
It turns out this template annotation was also mentioned in Psalm https://psalm.dev/docs/annotating_code/templated_annotations/#param-class-stringt

And League Container itself has a commit to improve this one https://github.com/thephpleague/container/commit/388e992884ab75632d93dbaf4d29ec024d64b964

RadoslavGeorgiev replied:
Amazing suggestion, @htdat! Implemented in https://github.com/Automattic/woocommerce-payments/pull/6862/commits/0766124aecb621f6d781d49e9e3ab76415010dd1

By the way, I never knew this existed in PHP.

htdat reviewed (approved):
Pre-approve PR. I also have the similar concern with `ExtendedContainer` as I also think it should be wrapped in the `tests` folder. But the answer from Rado makes sense. 

Instead of that, there are some minor inline comments. 

Pre-approve the PR.

RadoslavGeorgiev committed:
Using a dynamic return type

RadoslavGeorgiev committed:
Using Composer to autoload wcpay_get_container()

RadoslavGeorgiev committed:
Merge branch 'develop' into rpp/6678-dependency-container
